### PR TITLE
Sheet fixes

### LINKED ISSUE #8197: Unbuffered sheet exception in voxel renderer
Reported by @GraionDilach in IRC. Pinging @RoosterDragon.

```
Attacque Superior Mod at Version {DEV_VERSION}
Operating System: Windows (Microsoft Windows NT 6.1.7601 Service Pack 1)
Runtime Version: .NET CLR 4.0.30319.18444
Exception of type `System.InvalidOperationException`: This sheet is unbuffered. You cannot call CommitData on an unbuffered sheet. If you need to completely replace the texture data you should set data into the texture directly. If you need to make only small changes to the texture data consider creating a buffered sheet instead.
   at OpenRA.Graphics.Sheet.CommitData(Boolean releaseBuffer) in e:\OpenRA-dev\OpenRA.Game\Graphics\Sheet.cs:line 171
   at OpenRA.Graphics.Sheet.ReleaseBuffer() in e:\OpenRA-dev\OpenRA.Game\Graphics\Sheet.cs:line 163
   at OpenRA.Graphics.SheetBuilder.Allocate(Size imageSize, float2 spriteOffset) in e:\OpenRA-dev\OpenRA.Game\Graphics\SheetBuilder.cs:line 120
   at OpenRA.Graphics.VoxelRenderer.RenderAsync(WorldRenderer wr, IEnumerable`1 voxels, WRot camera, Single scale, Single[] groundNormal, WRot lightSource, Single[] lightAmbientColor, Single[] lightDiffuseColor, PaletteReference color, PaletteReference normals, PaletteReference shadowPalette) in e:\OpenRA-dev\OpenRA.Game\Graphics\VoxelRenderer.cs:line 164
   at OpenRA.Mods.Common.Graphics.VoxelRenderable.FinalizedVoxelRenderable..ctor(WorldRenderer wr, VoxelRenderable voxel)
   at OpenRA.Mods.Common.Graphics.VoxelRenderable.PrepareRender(WorldRenderer wr)
   at OpenRA.Graphics.WorldRenderer.<GenerateRenderables>b__6(IRenderable r) in e:\OpenRA-dev\OpenRA.Game\Graphics\WorldRenderer.cs:line 118
   at System.Linq.Enumerable.WhereSelectEnumerableIterator`2.MoveNext()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at OpenRA.Graphics.WorldRenderer.GenerateRenderables() in e:\OpenRA-dev\OpenRA.Game\Graphics\WorldRenderer.cs:line 117
   at OpenRA.Graphics.WorldRenderer.Draw() in e:\OpenRA-dev\OpenRA.Game\Graphics\WorldRenderer.cs:line 131
   at OpenRA.Game.RenderTick() in e:\OpenRA-dev\OpenRA.Game\Game.cs:line 501
   at OpenRA.Game.Loop() in e:\OpenRA-dev\OpenRA.Game\Game.cs:line 623
   at OpenRA.Game.Run() in e:\OpenRA-dev\OpenRA.Game\Game.cs:line 641
   at OpenRA.Program.Run(String[] args) in e:\OpenRA-dev\OpenRA.Game\Support\Program.cs:line 111
   at OpenRA.Program.Main(String[] args) in e:\OpenRA-dev\OpenRA.Game\Support\Program.cs:line 39
```

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/16226409-unbuffered-sheet-exception-in-voxel-renderer?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by RoosterDragon**
Fixes #8197 (disclaimer: untested).

The voxel renderer creates unbuffered sheets but still allocates sprites over them. This unusual use case caused the sheet builder to fall over as it always assumed buffered sheets would be created and thus need releasing.

Ultimately, releasing the buffer if the sheet is unbuffered is not a problem. You only need to get the exception when calling CommitData as that assumes you were writing to the buffer and now want that buffer committed to texture. If you try and commit and there was no buffer, that's a logic error you want to know about.
- If you call ReleaseBuffer and the buffer is already released, this now is a no-op rather than an exception.
- I have renamed CommitData to CommitBufferedData for a bit of extra clarity. This remains an exception if called when there is no buffer.
- The junk code doing some thread-safety in Sheet has been removed. It is not needed as the only cross-thread call was removed in #8103 because the thread-safety was half-baked anyway.

### CODE REVIEW

**Comment by GraionDilach** [comment]
Just popping in to say that it indeed fixes #8197. Tested.

File: OpenRA.Game/Graphics/Sheet.cs [hunk 1]
```diff
@@ -141,42 +128,32 @@ public Bitmap AsBitmap(TextureChannel channel, IPalette pal)
...
-					data = texture.GetData();
-				releaseBufferOnCommit = false;
-			}
+			if (data != null)
+				return;
+			if (texture == null)
+				data = new byte[4 * Size.Width * Size.Height];
+			else
+				data = texture.GetData();
+			releaseBufferOnCommit = false;
```
Comment (Phrohdoh) [inline OpenRA.Game/Graphics/Sheet.cs]: Newline before this line, please.
Sorry I can't comment on the logical changes here. :disappointed:

Comment (Mailaender) [inline OpenRA.Game/Graphics/Sheet.cs]: The real change without whitespace noise makes it easier to review https://github.com/OpenRA/OpenRA/pull/8200/files?w=1

**Comment by penev92** [comment]
OK :+1:

**Comment by penev92** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-%28bleed%29/_compare/979dd6b3963e01e8782fdd2dc42bc9a978b2da02...ba057fd6a9f5ef8a130d4811be7500a787a62a74)

**Pull request merged**